src/megachatapi_impl.cpp

```diff
@@ -131,37 +121,30 @@ void MegaChatApiImpl::loop()
 
         if (threadExit)
         {
-            // remove the MegaChatApiImpl that is being deleted
-            refsMutex.lock();
-            for (vector<MegaChatApiImpl*>::iterator it = megaChatApiRefs.begin(); it != megaChatApiRefs.end(); it++)
-            {
-                if (*it == this)
-                {
-                    megaChatApiRefs.erase(it);
-                    break;
-                }
-            }
-            sendPendingEvents();    // process any pending events in the queue
-            refsMutex.unlock();
+            //TODO: Check why there could be pending events here if the cleanup was correct
+            sendPendingEvents();
+
             sdkMutex.unlock();
             break;
         }
     }
 }
 
-void MegaChatApiImpl::megaApiPostMessage(void* msg)
-{
-    // Add the message to the queue of events
-    refsMutex.lock();
-    if (megaChatApiRefs.size())
+void MegaChatApiImpl::megaApiPostMessage(void* msg, void* ctx)
+{    
+    MegaChatApiImpl *megaChatApi = (MegaChatApiImpl *)ctx;
+    if (megaChatApi)
     {
-        megaChatApiRefs[0]->postMessage(msg);
+        megaChatApi->postMessage(msg);
     }
-    else    // no more instances running, only one left --> directly process messages
+    else
     {
+        // For compatibility with the QT example app,
+        // there are some marshallCall() without context
+        // that don't need to be marshalled using the
+        // intermediate layer
```

Comment:
Despite the Qt app doesn't use this method, there are a few marshal calls that do not attach any `appCtx` because it's not available at that point. Since the Qt app requires the marshaling, we allow those cases and directly execute the event.